src/corona_hakab_model/generation/geographic_circle.py

```diff
@@ -87,57 +88,69 @@ def create_social_circles_by_type(self, connection_type: ConnectionTypes, agents
+
+            # if the distribution is age dependent, fill with the appropriate age proportions
+            if connection_type in Non_Random_Age_Types:
+                circles.append(self.create_age_dependant_circle(connection_type, agents_for_type, circle_size))
+
+            else:
+                circle = SocialCircle(connection_type)
+                for _ in range(circle_size):
+                    agent = agents_for_type.pop()
+                    assert agent not in circle.agents
                     circle.add_agent(agent)
-                    agents_for_type.remove(agent)
-                    index += 1
-                    size_num_agents[size] -= 1
-           
-            # fill in the rest of the population 
-            while len(agents_for_type) > 0 and size_num_agents[size] > 0:
-                agent = agents_for_type.pop()
-                circle = circles[index % len(circles)]
-                circle.add_agent(agent)
-                index += 1
-                size_num_agents[size] -= 1
-            
-            self.connection_type_to_social_circles[connection_type].extend(circles)
-            self.all_social_circles.extend(circles)
+                circles.append(circle)
+
+        self.connection_type_to_social_circles[connection_type].extend(circles)
+        self.all_social_circles.extend(circles)
+
+    def create_age_dependant_circle(self, connection_type, agents_for_type, size):
+        circle = SocialCircle(connection_type)
+
+        # if circle size is abnormal, number of adults is taken from the closest smaller possible size
+        possible_sizes, _ = self.data_holder.circles_size_distribution_by_connection_type[connection_type]
+        if size in possible_sizes:
+            adult_type_distribution = self.data_holder.adult_distributions.get(connection_type)[size]
+        else:
+            approx_size = min([psize for psize in possible_sizes if psize < size], key=lambda el: abs(el - size))
+            adult_type_distribution = self.data_holder.adult_distributions.get(connection_type)[approx_size]
+
+        adults = [agent for agent in agents_for_type if agent.age > 18]
+        non_adults = [agent for agent in agents_for_type if agent.age <= 18]
+
+        adult_num = min(round(adult_type_distribution.rvs()), len(adults))
+        child_num = min(size - adult_num, len(non_adults))
+        for _ in range(adult_num):
+            agent = adults.pop()
+            assert agent not in circle.agents
+            circle.add_agent(agent)
+            agents_for_type.remove(agent)
+
+        for _ in range(child_num):
+            agent = non_adults.pop()
+            assert agent not in circle.agents
+            circle.add_agent(agent)
+            agents_for_type.remove(agent)
+
+        # if there is place left in the circle, fill it with agents:
+        while circle.agent_count < size:
+            agent = agents_for_type.pop()
+            assert agent not in circle.agents
```

Comment:
Yes, the assert was used to debug a specific issue. I haven't found a simpler way to select adults multiple adults, and remove them at once from adults and agents_for_type, so the loop would still be required

Reply:
Why not? You know how many are needed ( size - circle.agent_count) , you can take a cut out of agent_of_type use add_many.
Also pop sounds like a bad use, because you modify the outer list, a result which seems undesirable.

Reply:
The outer list should be modified because those agents were allocated to a circle of the current connection type, and should not be selected for similar circles again. I guess this can be accomplished more efficiently through list slicing, instead of this original implementation

Reply:
@denzo9  Check if it does so, what in actuall might happen with just slicing is that it will change your pointer

Reply:
I pushed a solution that seems to work, using slicing and the del keyword